cluster/swarm/cluster.go

```diff
@@ -26,24 +27,37 @@ type Cluster struct {
 	eventHandler cluster.EventHandler
 	engines      map[string]*cluster.Engine
 	scheduler    *scheduler.Scheduler
-	options      *cluster.Options
 	store        *state.Store
+
+	overcommitRatio float64
+	discovery       string
+	heartbeat       uint64
+	TLSConfig       *tls.Config
 }
 
 // NewCluster is exported
-func NewCluster(scheduler *scheduler.Scheduler, store *state.Store, options *cluster.Options) cluster.Cluster {
+func NewCluster(scheduler *scheduler.Scheduler, store *state.Store, TLSConfig *tls.Config, options cluster.DriverOpts) (cluster.Cluster, error) {
 	log.WithFields(log.Fields{"name": "swarm"}).Debug("Initializing cluster")
 
 	cluster := &Cluster{
-		engines:   make(map[string]*cluster.Engine),
-		scheduler: scheduler,
-		options:   options,
-		store:     store,
+		engines:         make(map[string]*cluster.Engine),
+		scheduler:       scheduler,
+		store:           store,
+		overcommitRatio: 0.05,
+		TLSConfig:       TLSConfig,
+	}
+
+	if ok, val := options.Float("swarm.overcommit"); ok {
+		cluster.overcommitRatio = val
+	}
+	_, cluster.discovery = options.String("swarm.discovery")
+	if _, cluster.heartbeat = options.Uint("swarm.heartbeat"); cluster.heartbeat < 1 {
+		return nil, errors.New("heartbeat should be an unsigned integer and greater than 0")
 	}
 
 	// get the list of entries from the discovery service
 	go func() {
-		d, err := discovery.New(options.Discovery, options.Heartbeat)
+		d, err := discovery.New(cluster.discovery, cluster.heartbeat)
```

Comment:
I agree, the thing is,  you need `swarm.heartbeat` to create the discovery, so it feels weird to parse a `swarm.` opt outside of swarm, to create the discovery and give it to swarm.